src/containerbuddy/etcd.go

```diff
@@ -0,0 +1,240 @@
+
+// Etcd is a service discovery backend for CoreOS etcd
+type Etcd struct {
+	Client client.Client
+	API    client.KeysAPI
+	Prefix string
+}
+
+// EtcdServiceNode is an instance of a service
+type EtcdServiceNode struct {
+	ID      string   `json:"id"`
+	Name    string   `json:"name"`
+	Address string   `json:"address"`
+	Port    int      `json:"port"`
+	Tags    []string `json:"tags,omitempty"`
+}
+
+// NewEtcdConfig creates a new service discovery backend for etcd
+func NewEtcdConfig(config map[string]interface{}) Etcd {
+	etcd := Etcd{
+		Prefix: "/containerbuddy",
+	}
+	etcdConfig := client.Config{}
+	switch endpoints := config["endpoints"].(type) {
+	case string:
+		etcdConfig.Endpoints = []string{endpoints}
+	case []string:
+		etcdConfig.Endpoints = endpoints
+	default:
+		log.Fatal("Must provide etcd endpoints")
+	}
+
+	prefix, ok := config["prefix"].(string)
+	if ok {
+		etcd.Prefix = prefix
+	}
+
+	etcdClient, err := client.New(etcdConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	etcd.Client = etcdClient
+	etcd.API = client.NewKeysAPI(etcdClient)
+	return etcd
+}
+
+// Deregister removes this instance from the registry
+func (c Etcd) Deregister(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (c Etcd) MarkForMaintenance(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// SendHeartbeat refreshes the TTL of this associated etcd node
+func (c Etcd) SendHeartbeat(service *ServiceConfig) {
+	if c.checkServiceExists(service) {
+		c.updateServiceTTL(service)
+	} else {
+		log.Printf("Service not registered, registering...")
+		if err := c.registerService(service); err != nil {
+			log.Printf("Error registering service %s: %s", service.Name, err)
+		}
+
+	}
+}
+
+// CheckForUpstreamChanges checks another etcd node for changes
+func (c Etcd) CheckForUpstreamChanges(backend *BackendConfig) bool {
+	return c.checkHealth(backend)
+}
+
+func (c Etcd) getNodeKey(service *ServiceConfig) string {
+	return fmt.Sprintf("%s/%s/%s", c.Prefix, service.Name, service.ID)
+}
+
+func (c Etcd) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", c.Prefix, appName)
+}
+
+var etcdUpstreams = make(map[string][]EtcdServiceNode)
+
+func (c Etcd) checkHealth(backend *BackendConfig) bool {
+	services, err := c.getServices(backend.Name)
+	if err != nil {
+		if _, ok := err.(client.Error); !ok {
+			log.Printf("Failed to query %v: %s", backend.Name, err)
+		}
+		return false
+	}
+	didChange := etcdCompareForChange(etcdUpstreams[backend.Name], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		etcdUpstreams[backend.Name] = services
+	}
+	return didChange
+}
+
+func (c Etcd) checkServiceExists(service *ServiceConfig) bool {
+	key := c.getNodeKey(service)
+	_, err := c.API.Get(context.Background(), key, nil)
+	if err != nil {
```

Comment:
The golang-idiomatic way to do this test if you don't need the return values to live outside the scope of the `if` block is:

```go
if _, err := c.API.Get(context.Background(), key, nil); err != nil {
    // stuff
}
```